server/src/main/java/org/elasticsearch/action/admin/indices/resolve/ResolveClusterActionResponse.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.admin.indices.resolve;
+
+import org.elasticsearch.Build;
+import org.elasticsearch.TransportVersions;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.transport.RemoteClusterAware;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+
+public class ResolveClusterActionResponse extends ActionResponse implements ToXContentObject {
+
+    private static final ParseField CONNECTED_FIELD = new ParseField("connected");
+    private static final ParseField SKIP_UNAVAILABLE_FIELD = new ParseField("skip_unavailable");
+    private static final ParseField MATCHING_INDICES_FIELD = new ParseField("matching_indices");
+    private static final ParseField ES_VERSION_FIELD = new ParseField("version");
+    private static final ParseField ERROR_FIELD = new ParseField("error");
+
+    private final Map<String, ResolveClusterInfo> infoMap;
+
+    public ResolveClusterActionResponse(Map<String, ResolveClusterInfo> infoMap) {
+        this.infoMap = infoMap;
+    }
+
+    public ResolveClusterActionResponse(StreamInput in) throws IOException {
+        super(in);
+        if (in.getTransportVersion().before(TransportVersions.RESOLVE_CLUSTER_ENDPOINT_ADDED)) {
+            throw new UnsupportedOperationException(
```

Comment:
This PR [was requested](https://github.com/elastic/elasticsearch/pull/102726#discussion_r1417759444) to start doing this for a brand new Writable even though I don't see that Elasticsearch has ever done this before. The recommended approach was to throw an exception: https://github.com/elastic/elasticsearch/pull/102726#discussion_r1432363148

There is also a test in `TransportResolveClusterActionTests` that gets this error message.

Reply:
It makes sense to do this when we write, to avoid serializing what cannot be deserialized on the other end and be upfront on the fact that this API is not supported across all nodes, hence it cannot work. I don't understand how production code will exercise this scenario when reading, maybe I am missing something. We are in the constructors of the new response, in the version that introduces the class. If the version is lower, the remote node cannot possibly have sent this response class? Am I missing something?

Reply:
I see what you mean. I removed the version checks from the StreamInput based methods but left them in the write methods. The older version cluster detection still works as before, so the writing check is indeed what is needed. Fixed in next push.